File: backend/app/schemas/meals.py
# app/schemas/meals.py
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field, field_validator


class MealCreate(BaseModel):
    meal_text: str = Field(min_length=1, max_length=2000)
    segment: str
    date: str
    calories: float = Field(ge=0)
    protein: float = Field(default=0.0, ge=0)
    carbs: float = Field(default=0.0, ge=0)
    fibre: float = Field(default=0.0, ge=0)
    fats: float = Field(default=0.0, ge=0)
    confidence: Optional[float] = None

    @field_validator("confidence", mode="before")
    @classmethod
    def _coerce_confidence(cls, v):
        if v is None or isinstance(v, (int, float)):
            return v
        try:
            return float(v)
        except (TypeError, ValueError):
            mapping = {
                "low": 0.3,
                "medium": 0.6,
                "med": 0.6,
                "moderate": 0.6,
                "high": 0.85,
                "very high": 0.95,
            }
            return mapping.get(str(v).strip().lower(), 0.5)


class MealOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    meal_text: str
    segment: str
    date: str
    calories: float
    protein: float
    carbs: float
    fibre: float
    fats: float
    confidence: Optional[float] = None
    created_at: str

File: frontend/src/models/meals.ts
// src/models/meals.ts
export type MealSegmentType = "Breakfast" | "Lunch" | "Dinner" | "Snacks";

export interface Meal {
  id: string;
  meal_text: string;
  segment: MealSegmentType;
  date: string;
  calories: number;
  protein: number;
  carbs: number;
  fibre: number;
  fats: number;
  confidence?: number | null;
  created_at: string;
}

export interface MealCreateInput {
  meal_text: string;
  segment: string;
  date: string;
  calories: number;
  protein: number;
  carbs: number;
  fibre: number;
  fats: number;
  confidence?: number | null;
}
